apps/language_server/lib/language_server/providers/folding_range/line.ex

```diff
@@ -0,0 +1,42 @@
+defmodule ElixirLS.LanguageServer.Providers.FoldingRange.Line do
+  @moduledoc """
+  FoldingRange helpers for lines.
+  """
+
+  alias ElixirLS.LanguageServer.SourceFile
+
+  @type cell :: {non_neg_integer(), non_neg_integer() | nil}
+  @type t :: {String.t(), cell(), String.t()}
+
+  @spec format_string(String.t()) :: [cell()]
+  def format_string(text) do
+    text
+    |> SourceFile.lines()
+    |> embellish_lines_with_metadata()
+  end
+
+  # If we think of the code text as a grid, this function finds the cells whose
+  # columns are the start of each row (line).
+  # Empty rows are represented as {row, nil}.
+  # We also grab the first character for convenience elsewhere.
+  @spec embellish_lines_with_metadata([String.t()]) :: [t()]
+  defp embellish_lines_with_metadata(lines) do
+    lines
+    |> Enum.with_index()
+    |> Enum.map(fn {line, row} ->
+      full_length = line |> SourceFile.line_length_utf16()
+      trimmed = line |> String.trim_leading()
+      trimmed_length = trimmed |> SourceFile.line_length_utf16()
+      first = trimmed |> String.first()
+
+      col =
+        if {full_length, trimmed_length} == {0, 0} do
```

Comment:
Good question! I went back and forth about this. Here's my reasoning:

Since this pass is driven by whitespace and ignores the syntax, and since this pass has the lowest priority (it's the fallback in case I failed to cover something in the other passes), I'm inclined to honor a whitespace-only line.

But I'm happy to go in the other direction if that's what we want! This PR is full of little implicit choices like that.

Reply:
That makes sense. I really don't have a strong opinion on this. Anyway elixir-ls automatically removes whitespace from empty lines if you use the formatter so I guess for most people it won't even matter.